File: Data Structures & Algorithms/foreign-dictionary/submission-1.cpp
class Solution {
private:
    unordered_map<char, vector<char>> adjList;
    unordered_map<char, bool> visited;
    unordered_map<char, bool> pathVisited;
    vector<char> res;

    bool isCycle(char ch){
        if(visited[ch] && pathVisited[ch]){
            return true;
        }

        if(visited[ch]){
            return false;
        }

        visited[ch] = true;
        pathVisited[ch] = true;

        for(char next: adjList[ch]){
            if(isCycle(next)){
                return true;
            }
        }

        res.push_back(ch);
        pathVisited[ch] = false;
        return false;
    }

public:
    string foreignDictionary(vector<string>& words) {
        for(string& word: words){
            for(char ch: word){
                adjList[ch];
            }
        }
        
        for(int i=0; i<words.size()-1; i++){
            string& a = words[i];
            string& b = words[i+1];

            int minLen = min(a.length(), b.length());

            if((a.substr(0,minLen) == b.substr(0,minLen))
                && (a.length() > b.length())){
                    return "";
            }



            for(int i=0; i<a.length(); i++){
                if(a[i] != b[i]){
                    adjList[a[i]].push_back(b[i]);
                    break;
                }
            }
        }

        for(auto& pair: adjList){
            if(isCycle(pair.first)){
                return "";
            }
        }

        reverse(res.begin(),res.end());

        string order = "";
        for(char ch: res){
            order += ch;
        }

        return order;
    }
};

File: Data Structures & Algorithms/spiral-matrix/submission-1.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m,n = len(matrix), len(matrix[0])
        t,b,l,r = 0,m-1,0,n-1

        res=[]
        while t<=b and l<=r:
            for j in range(l,r+1):
                res.append(matrix[t][j])
            t+=1
            for i in range(t,b+1):
                res.append(matrix[i][r])
            r-=1
            if t>b or l>r:
                break
            for j in range(r,l-1,-1):
                res.append(matrix[b][j])
            b-=1
            for i in range(b,t-1,-1):
                res.append(matrix[i][l])
            l+=1
        
        return res
            

        